packages/resolvers/src/resolution_result_cache_resolver.rs

```diff
@@ -0,0 +1,128 @@
+        uri_resolver::UriResolver
+    },
+    error::Error
+};
+use crate::cache::basic_resolution_result_cache::BasicResolutionResultCache;
+use crate::cache::resolution_result_cache::ResolutionResultCache;
+use crate::uri_resolver_aggregator::UriResolverAggregator;
+
+/// A URI resolver that uses a cache to store and retrieve wrappers that pass through.
+pub struct ResolutionResultCacheResolver {
+    resolver: Arc<dyn UriResolver>,
+    cache: Mutex<Box<dyn ResolutionResultCache>>,
+}
+
+impl ResolutionResultCacheResolver {
+    /// Creates a new `ResolutionResultCacheResolver`.
+    ///
+    /// # Arguments
+    ///
+    /// * `resolver` - The `UriResolver` to use when resolving URIs.
+    /// * `cache` - The cache to store and retrieve resolved URIs.
+    ///
+    /// # Returns
+    ///
+    /// * A new `ResolutionResultCacheResolver`.
+    pub fn new(resolver: Arc<dyn UriResolver>, cache: Mutex<Box<dyn ResolutionResultCache>>) -> ResolutionResultCacheResolver {
+        ResolutionResultCacheResolver { resolver, cache }
+    }
+
+    fn cache_resolution_path(&self, resolution_context: Arc<Mutex<UriResolutionContext>>, result: Arc<Result<UriPackageOrWrapper, Error>>) {
+        let resolution_path = resolution_context.lock().unwrap().get_resolution_path();
+        for uri in resolution_path {
+            self.cache.lock().unwrap().set(uri, result.clone());
+        }
+    }
+}
+
+impl UriResolver for ResolutionResultCacheResolver {
+    /// Tries to resolve the given URI using a cache and returns the result.
+    ///
+    /// # Arguments
+    ///
+    /// * `uri` - The URI to resolve.
+    /// * `invoker` - The invoker of the resolution.
+    /// * `resolution_context` - The context for the resolution.
+    ///
+    /// # Returns
+    ///
+    /// * A `Result` containing the resolved `UriPackageOrWrapper` on success, or an exception on failure.
+    fn try_resolve_uri(
+        &self,
+        uri: &Uri,
+        invoker: Arc<dyn Invoker>,
+        resolution_context: Arc<Mutex<UriResolutionContext>>,
+    ) -> Result<UriPackageOrWrapper, Error> {
+        if let Some(cache_result) = self.cache.lock().unwrap().get(uri) {
+            let result = cache_result.clone().deref().clone();
+            resolution_context.lock().unwrap().track_step(
+                UriResolutionStep {
+                    source_uri: uri.clone(),
+                    result: result.clone(),
+                    sub_history: None,
+                    description: Some("ResolutionResultCacheResolver (Cache)".to_string()),
+                }
+            );
+            return result;
+        }
+
+        let sub_context = resolution_context.lock().unwrap().create_sub_history_context();
+        let sub_context = Arc::new(Mutex::new(sub_context));
+        let result = self.resolver.try_resolve_uri(uri, invoker.clone(), sub_context.clone());
+
+        if result.is_ok() {
+            self.cache_resolution_path(sub_context.clone(), Arc::from(result.clone()));
```

Comment:
We should only cache the current uri here, not the whole path.
That way we keep the resolution path the same for cached and non cached wraps.
(This fixes issues with env which is the main reason we have this resolver)